File: server/server/TravelAgencyServicesImplementation.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using model;
using persistence.reservationRepo;
using persistence.tripRepo;
using persistence.userRepo;
using services;

namespace server.server
{

    public class TravelAgencyServicesImplementation : ITravelAgencyServices
    {
        private IUserRepository _userRepository;
        private ITripRepository _tripRepository;
        private IReservationRepository _reservationRepository;
        private readonly IDictionary<string, ITravelAgencyObserver> _loggedClients;

        public TravelAgencyServicesImplementation(IUserRepository userRepository, ITripRepository tripRepository, IReservationRepository reservationRepository)
        {
            _userRepository = userRepository;
            _tripRepository = tripRepository;
            _reservationRepository = reservationRepository;
            _loggedClients = new Dictionary<string, ITravelAgencyObserver>();
        }

        #nullable enable
        public void Login(User user, ITravelAgencyObserver client)
        {
            User? targetUser = _userRepository.FindUserByUsername(user.Username);
            if (targetUser != null)
            {
                if (_loggedClients.ContainsKey(user.Username))
                {
                    throw new TravelAgencyException("User already logged!");
                    return;
                }

                if (user.Password == targetUser.Password)
                {
                    _loggedClients[user.Username] = client;
                    return;
                }
            }
            
            throw new TravelAgencyException("Authentication failed!");
        }

        public void ReserveTrip(Reservation reservation)
        {
            User? user = _userRepository.FindUserByUsername(reservation.AgencyUser.Username);
            if (user == null)
            {
                throw new TravelAgencyException("User unknown!");
            }

            reservation.AgencyUser.Id = user.Id;

            Reservation? newReservation = _reservationRepository.Save(reservation);
            if (newReservation != null)
            {
                throw new TravelAgencyException("Error while trying to save reservation");
            }

            Trip trip = reservation.CurrentTrip;
            trip.AvailableSeats = trip.AvailableSeats - reservation.NumberOfTickets;
            Trip? updatedTrip = _tripRepository.Update(trip);

            if (updatedTrip != null)
            {
                throw new TravelAgencyException("Error while trying to update trip");
            }

            NotifyUsersLoggedIn(reservation);
        }

        public Trip[] GetTrips()
        {
            IEnumerable<Trip> trips = _tripRepository.FindAll();
            return trips.ToArray();
        }
        
        public Trip[] GetSearchedTrips(string landmark, int startHour, int endHour)
        {
            IEnumerable<Trip> searchedTrips = _tripRepository.FindBetweenHoursHavingLandmark(landmark, startHour, endHour);
            return searchedTrips.ToArray();
        }

        public void Logout(User user, ITravelAgencyObserver client)
        {
            ITravelAgencyObserver loggedUser = _loggedClients[user.Username];
            if (loggedUser == null)
            {
                throw new TravelAgencyException("User " + user.Username + " is not logged in!");
            }

            _loggedClients.Remove(user.Username);
        }

        private void NotifyUsersLoggedIn(Reservation reservation)
        {
            foreach (ITravelAgencyObserver client in _loggedClients.Values)
            {
                Task.Run(() => client.TripReserved(reservation));
            }
        }
    }
}

File: Java/TravelAgencyServer/src/main/java/travelAgency/server/TravelAgencyServicesImplementation.java
package travelAgency.server;

import at.favre.lib.crypto.bcrypt.BCrypt;
import travelAgency.model.Reservation;
import travelAgency.model.Trip;
import travelAgency.model.User;
import travelAgency.persistence.ReservationRepository;
import travelAgency.persistence.TripRepository;
import travelAgency.persistence.UserRepository;
import travelAgency.services.ITravelAgencyObserver;
import travelAgency.services.ITravelAgencyServices;
import travelAgency.services.TravelAgencyException;

import java.io.Console;
import java.lang.constant.Constable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class TravelAgencyServicesImplementation implements ITravelAgencyServices {
    private final int defaulThreadsNo = 5;
    private UserRepository userRepository;
    private TripRepository tripRepository;
    private ReservationRepository reservationRepository;
    private Map<String, ITravelAgencyObserver> loggedUsers;

    public TravelAgencyServicesImplementation(UserRepository userRepository, TripRepository tripRepository, ReservationRepository reservationRepository) {
        this.userRepository = userRepository;
        this.tripRepository = tripRepository;
        this.reservationRepository = reservationRepository;

        //userRepository.save(new User("razvan", "razvan"));
        loggedUsers = new ConcurrentHashMap<>();
    }

    private String hashPassword(String password) {
        // Hash a password
        return BCrypt.withDefaults().hashToString(12, password.toCharArray());
    }

    private void notifyUsersLoggedIn(Reservation reservation) throws TravelAgencyException {
        ExecutorService executor = Executors.newFixedThreadPool(defaulThreadsNo);
        for(ITravelAgencyObserver travelAgencyClient : loggedUsers.values()) {
            executor.execute(() -> {
                try {
                    travelAgencyClient.tripReserved(reservation);
                } catch (TravelAgencyException e) {
                    System.err.println("Error notifying user " + e);
                }
            });
        }
    }

    @Override
    public synchronized void login(User user, ITravelAgencyObserver client) throws TravelAgencyException {
        Optional<User> nUser = userRepository.findUserByUsername(user.getUsername());

        if(nUser.isPresent()) {
            if(nUser.get().getPassword().equals(user.getPassword())) {
                if(loggedUsers.get(user.getUsername()) != null) {
                    throw new TravelAgencyException("User already logged in");
                } else {
                    loggedUsers.put(user.getUsername(), client);
                }
            } else {
                throw new TravelAgencyException("Authentication failed");
            }
        }
        else {
            throw new TravelAgencyException("Authentication failed");
        }
    }

    @Override
    public void reserveTrip(Reservation reservation) throws TravelAgencyException {
        User user = reservation.getAgencyUser();
        Optional<Reservation> nReservation = reservationRepository.save(reservation);

        Trip trip = reservation.getTrip();
        trip.setAvailableSeats(trip.getAvailableSeats() - reservation.getNumberOfTickets());
        Optional<Trip> tripU = tripRepository.update(trip.getId(), trip);

        if(nReservation.isPresent()) {
            throw new TravelAgencyException("Error while saving reservation");
        }

        if(tripU.isPresent()) {
            throw new TravelAgencyException("Error while trying to update trip");
        }

        notifyUsersLoggedIn(reservation);
    }

    @Override
    public void logout(User user, ITravelAgencyObserver client) throws TravelAgencyException {
        ITravelAgencyObserver localUser = loggedUsers.remove(user.getUsername());
        if(localUser == null) {
            throw new TravelAgencyException("User is not logged in");
        }
    }

    @Override
    public Trip[] getTrips() throws TravelAgencyException {
        Iterable<Trip> tripsIterable = tripRepository.findAll();

        List<Trip> tripList = new ArrayList<>();
        tripsIterable.forEach(tripList::add);

        return tripList.toArray(new Trip[0]);
    }

    @Override
    public Trip[] getSearchedTrips(String landmark, int startHour, int endHour) throws TravelAgencyException {
        List<Trip> tripIterable = tripRepository.findBetweenHoursHavingLandmark(startHour, endHour, landmark);
        return tripIterable.toArray(new Trip[0]);
    }
}
